### PR TITLE
expose drop_in_place as ptr::drop_in_place

### PR DESCRIPTION
**Pull request by Gankra**
This intrinsic is necessary to drop DSTs and should generally be the way to drop memory before manually deallocating it. As such, it should be properly exposed by the standard library.

### CODE REVIEW

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -194,7 +194,25 @@ extern "rust-intrinsic" {
...
+    ///
+    /// * It is friendlier to the optimizer to do this over `ptr::read` when
+    ///   dropping manually allocated memory (e.g. when writing Box/Rc/Vec),
+    ///   as the compiler doesn't need to prove that it's sound to elide the
+    ///   copy.
+    ///
+    /// # Undefined Behaviour
+    ///
+    ///   This has all the same safety problems as `ptr::read` with respect to
+    ///   invalid pointers, types, and double drops.
```
Comment (alexcrichton) [inline src/libcore/intrinsics.rs]: extra tabbing in these comments

**Comment by Gankro** [comment]
The outstanding issue for this is ptr::drop_in_place vs mem::drop_in_place and whether it should take a *mut or an &mut.

mem seems more consistent with mem::drop, but ptr is strictly more flexible (&mut coerces to *mut).

**Comment by huonw** [comment]
(This will need a tracking issue and an `issue = "..."` to land.)

File: src/libcore/ptr.rs [hunk 2]
```diff
@@ -39,6 +39,9 @@ pub use intrinsics::copy;
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use intrinsics::write_bytes;
 
+#[unstable(feature = "drop_in_place", reason = "just exposed, needs approval")]
```
Comment (alexcrichton) [inline src/libcore/ptr.rs]: In addition to what @huonw said about this needing `issue=` now this'll also need to be applied to the definition in the intrinsics module itself to work (we don't look at reexports themselves)

**Comment by alexcrichton** [comment]
r=me with @huonw's and my tweaks to the stability attribute

**Comment by Gankro** [comment]
@bors r=alexcrichton

(making the feature gate actually do a thing broke stuff)

**Comment by Gankro** [comment]
fsk how did that conflict even happen

**Comment by Gankro** [comment]
@bors r=alexcrichton

(This branch was super old, that's how!)

**Comment by alexcrichton** [comment]
(looks like travis is unhappy)

**Comment by Gankro** [comment]
augh run-pass test using the intrinsic

**Comment by Gankro** [comment]
(actually did a local make-check)

**Comment by apasel422** [comment]
What's the status of this?

**Comment by Gankro** [comment]
@apasel422 I'm kinda AWOL; feel free to rebase if you want this to land!

**Comment by apasel422** [comment]
Will do!

**Pull request closed without merging**